File: clients/go/acteon/queues_test.go
package acteon

// Go SDK task-queue surface tests.
//
// Live HTTP tests would need a running Acteon instance with the
// queue feature enabled; these tests exercise the wire surface of
// `queues.go` via an `httptest.Server` (reusing the capturing
// server from the A2A tests). The contract under test: URLs and
// methods match the server routes, request bodies snake-case and
// drop optional fields, task responses round-trip, and structured
// errors surface as `*APIError`.

import (
	"context"
	"encoding/json"
	"strings"
	"testing"
	"time"
)

// taskWire builds a minimal wire-form task body the mock server
// returns, with overrides merged on top.
func taskWire(overrides map[string]any) map[string]any {
	body := map[string]any{
		"task_id":      "t-1",
		"queue":        "q",
		"action_type":  "send_email",
		"payload":      map[string]any{"to": "a@b.c"},
		"status":       "pending",
		"attempt":      0,
		"max_attempts": 3,
		"created_at":   "2026-06-10T00:00:00Z",
		"updated_at":   "2026-06-10T00:00:00Z",
	}
	for k, v := range overrides {
		body[k] = v
	}
	return body
}

func TestEnqueueTaskURLAndBody(t *testing.T) {
	url, captured, teardown := newCapturingServer(t, 201, taskWire(nil))
	defer teardown()
	c := NewClient(url, WithAPIKey("k"))

	task, err := c.EnqueueTask(context.Background(), "q", &EnqueueTaskRequest{
		Namespace:   "ns",
		Tenant:      "tnt",
		ActionType:  "send_email",
		Payload:     map[string]any{"to": "a@b.c"},
		MaxAttempts: ptr(5),
	})
	if err != nil {
		t.Fatalf("enqueue: %v", err)
	}
	if captured.method != "POST" {
		t.Errorf("method: got %s", captured.method)
	}
	if captured.path != "/v1/queues/q/tasks" {
		t.Errorf("path: got %s", captured.path)
	}
	if got := captured.headers.Get("Authorization"); got != "Bearer k" {
		t.Errorf("Authorization: got %q", got)
	}
	s := string(captured.body)
	for _, want := range []string{
		`"namespace":"ns"`,
		`"tenant":"tnt"`,
		`"action_type":"send_email"`,
		`"max_attempts":5`,
	} {
		if !strings.Contains(s, want) {
			t.Errorf("expected %s in body; got %s", want, s)
		}
	}
	if task.TaskID != "t-1" || task.Status != TaskStatusPending {
		t.Errorf("task: got %+v", task)
	}
}

func TestEnqueueTaskOmitsMaxAttemptsWhenNil(t *testing.T) {
	url, captured, teardown := newCapturingServer(t, 201, taskWire(nil))
	defer teardown()
	c := NewClient(url)

	if _, err := c.EnqueueTask(context.Background(), "q", &EnqueueTaskRequest{
		Namespace:  "ns",
		Tenant:     "tnt",
		ActionType: "send_email",
		Payload:    map[string]any{},
	}); err != nil {
		t.Fatalf("enqueue: %v", err)
	}
	if strings.Contains(string(captured.body), "max_attempts") {
		t.Errorf("max_attempts should be omitted when nil; got %s", captured.body)
	}
}

func TestPollTasksURLBodyAndLeaseFields(t *testing.T) {
	leased := taskWire(map[string]any{
		"status":           "leased",
		"attempt":          1,
		"lease_token":      "lease-abc",
		"lease_expires_at": "2026-06-10T00:01:00Z",
	})
	url, captured, teardown := newCapturingServer(t, 200, map[string]any{
		"tasks": []map[string]any{leased},
	})
	defer teardown()
	c := NewClient(url)

	tasks, err := c.PollTasks(context.Background(), "q", &PollTasksRequest{
		Namespace:    "ns",
		Tenant:       "tnt",
		MaxTasks:     ptr(4),
		LeaseSeconds: ptr(120),
		WorkerID:     "w-1",
	})
	if err != nil {
		t.Fatalf("poll: %v", err)
	}
	if captured.method != "POST" || captured.path != "/v1/queues/q/poll" {
		t.Errorf("request: got %s %s", captured.method, captured.path)
	}
	s := string(captured.body)
	for _, want := range []string{
		`"max_tasks":4`,
		`"lease_seconds":120`,
		`"worker_id":"w-1"`,
	} {
		if !strings.Contains(s, want) {
			t.Errorf("expected %s in body; got %s", want, s)
		}
	}
	if len(tasks) != 1 {
		t.Fatalf("tasks: got %d", len(tasks))
	}
	got := tasks[0]
	if got.Status != TaskStatusLeased || got.LeaseToken != "lease-abc" {
		t.Errorf("leased task: got %+v", got)
	}
	if got.LeaseExpiresAt == nil || !got.LeaseExpiresAt.Equal(time.Date(2026, 6, 10, 0, 1, 0, 0, time.UTC)) {
		t.Errorf("lease_expires_at: got %v", got.LeaseExpiresAt)
	}
	var payload map[string]any
	if err := json.Unmarshal(got.Payload, &payload); err != nil {
		t.Fatalf("payload unmarshal: %v", err)
	}
	if payload["to"] != "a@b.c" {
		t.Errorf("payload: got %v", payload)
	}
}

func TestPollTasksEmptyQueueReturnsNoTasks(t *testing.T) {
	url, _, teardown := newCapturingServer(t, 200, map[string]any{"tasks": []map[string]any{}})
	defer teardown()
	c := NewClient(url)

	tasks, err := c.PollTasks(context.Background(), "q", &PollTasksRequest{Namespace: "ns", Tenant: "tnt"})
	if err != nil {
		t.Fatalf("poll: %v", err)
	}
	if len(tasks) != 0 {
		t.Errorf("tasks: got %d", len(tasks))
	}
}

func TestHeartbeatTaskURLAndBody(t *testing.T) {
	url, captured, teardown := newCapturingServer(t, 200, taskWire(map[string]any{"status": "leased"}))
	defer teardown()
	c := NewClient(url)

	if _, err := c.HeartbeatTask(context.Background(), "t-1", &HeartbeatTaskRequest{
		Namespace:     "ns",
		Tenant:        "tnt",
		LeaseToken:    "lease-abc",
		ExtendSeconds: ptr(90),
	}); err != nil {
		t.Fatalf("heartbeat: %v", err)
	}
	if captured.method != "POST" || captured.path != "/v1/queues/tasks/t-1/heartbeat" {
		t.Errorf("request: got %s %s", captured.method, captured.path)
	}
	s := string(captured.body)
	if !strings.Contains(s, `"lease_token":"lease-abc"`) || !strings.Contains(s, `"extend_seconds":90`) {
		t.Errorf("body: got %s", s)
	}
}

func TestCompleteTaskURLAndBody(t *testing.T) {
	url, captured, teardown := newCapturingServer(t, 200, taskWire(map[string]any{"status": "completed"}))
	defer teardown()
	c := NewClient(url)

	task, err := c.CompleteTask(context.Background(), "t-1", &CompleteTaskRequest{
		Namespace:  "ns",
		Tenant:     "tnt",
		LeaseToken: "lease-abc",
		Result:     map[string]any{"sent": true},
	})
	if err != nil {
		t.Fatalf("complete: %v", err)
	}
	if captured.method != "POST" || captured.path != "/v1/queues/tasks/t-1/complete" {
		t.Errorf("request: got %s %s", captured.method, captured.path)
	}
	if !strings.Contains(string(captured.body), `"result":{"sent":true}`) {
		t.Errorf("body: got %s", captured.body)
	}
	if task.Status != TaskStatusCompleted {
		t.Errorf("status: got %s", task.Status)
	}
}

func TestFailTaskURLAndBody(t *testing.T) {
	url, captured, teardown := newCapturingServer(t, 200, taskWire(map[string]any{"status": "failed"}))
	defer teardown()
	c := NewClient(url)

	if _, err := c.FailTask(context.Background(), "t-1", &FailTaskRequest{
		Namespace:  "ns",
		Tenant:     "tnt",
		LeaseToken: "lease-abc",
		Error:      "boom",
		Retryable:  false,
	}); err != nil {
		t.Fatalf("fail: %v", err)
	}
	if captured.method != "POST" || captured.path != "/v1/queues/tasks/t-1/fail" {
		t.Errorf("request: got %s %s", captured.method, captured.path)
	}
	s := string(captured.body)
	if !strings.Contains(s, `"error":"boom"`) || !strings.Contains(s, `"retryable":false`) {
		t.Errorf("body: got %s", s)
	}
}

func TestGetTaskURLAndQueryParams(t *testing.T) {
	url, captured, teardown := newCapturingServer(t, 200, taskWire(nil))
	defer teardown()
	c := NewClient(url)

	task, err := c.GetTask(context.Background(), "t-1", "ns", "tnt")
	if err != nil {
		t.Fatalf("get: %v", err)
	}
	if captured.method != "GET" || !strings.HasPrefix(captured.path, "/v1/queues/tasks/t-1") {
		t.Errorf("request: got %s %s", captured.method, captured.path)
	}
	if task.TaskID != "t-1" {
		t.Errorf("task: got %+v", task)
	}
}

func TestGetTaskNotFoundReturnsNil(t *testing.T) {
	url, _, teardown := newCapturingServer(t, 404, map[string]any{"error": "task not found"})
	defer teardown()
	c := NewClient(url)

	task, err := c.GetTask(context.Background(), "missing", "ns", "tnt")
	if err != nil {
		t.Fatalf("get: %v", err)
	}
	if task != nil {
		t.Errorf("task must be nil on 404: got %+v", task)
	}
}

func TestListTasksStatusFilter(t *testing.T) {
	url, captured, teardown := newCapturingServer(t, 200, map[string]any{
		"tasks": []map[string]any{taskWire(nil)},
	})
	defer teardown()
	c := NewClient(url)

	tasks, err := c.ListTasks(context.Background(), "q", "ns", "tnt", TaskStatusPending)
	if err != nil {
		t.Fatalf("list: %v", err)
	}
	if captured.method != "GET" || !strings.HasPrefix(captured.path, "/v1/queues/q/tasks") {
		t.Errorf("request: got %s %s", captured.method, captured.path)
	}
	if len(tasks) != 1 {
		t.Errorf("tasks: got %d", len(tasks))
	}
}

func TestQueuePathSegmentsArePercentEncoded(t *testing.T) {
	url, captured, teardown := newCapturingServer(t, 200, map[string]any{"tasks": []map[string]any{}})
	defer teardown()
	c := NewClient(url)

	// A queue name with a slash must be percent-encoded so it
	// cannot leak into additional path components.
	if _, err := c.PollTasks(context.Background(), "q/escape", &PollTasksRequest{
		Namespace: "ns",
		Tenant:    "tnt",
	}); err != nil {
		t.Fatalf("poll: %v", err)
	}
	if !strings.Contains(captured.path, "/q%2Fescape/") {
		t.Errorf("path must percent-encode slash: got %s", captured.path)
	}
}

func TestQueueErrorSurfacesAsAPIError(t *testing.T) {
	url, _, teardown := newCapturingServer(t, 409, map[string]any{
		"error": "lease token mismatch",
	})
	defer teardown()
	c := NewClient(url)

	_, err := c.CompleteTask(context.Background(), "t-1", &CompleteTaskRequest{
		Namespace:  "ns",
		Tenant:     "tnt",
		LeaseToken: "stale",
		Result:     nil,
	})
	if err == nil {
		t.Fatalf("expected error on 409")
	}
	apiErr, ok := err.(*APIError)
	if !ok {
		t.Fatalf("expected *APIError, got %T: %v", err, err)
	}
	if !strings.Contains(apiErr.Message, "lease token mismatch") {
		t.Errorf("message: got %q", apiErr.Message)
	}
	if apiErr.IsRetryable() {
		t.Errorf("409 must not be retryable")
	}
}

func TestQueueServerErrorIsRetryable(t *testing.T) {
	url, _, teardown := newCapturingServer(t, 500, map[string]any{"error": "backend down"})
	defer teardown()
	c := NewClient(url)

	_, err := c.PollTasks(context.Background(), "q", &PollTasksRequest{Namespace: "ns", Tenant: "tnt"})
	if err == nil {
		t.Fatalf("expected error on 500")
	}
	apiErr, ok := err.(*APIError)
	if !ok {
		t.Fatalf("expected *APIError, got %T: %v", err, err)
	}
	if !apiErr.IsRetryable() {
		t.Errorf("500 must be retryable")
	}
}

File: clients/nodejs/src/queues.test.ts
/**
 * Worker task queue surface tests.
 *
 * Live HTTP tests would need a running Acteon instance with task
 * queues enabled; these exercise serde (request body builders +
 * response parsers) and the `ActeonClient` queue methods via a
 * captured `fetch` stub. The contract under test: every wire field
 * round-trips, optional fields drop cleanly, and URLs/bodies match
 * the `/v1/queues` lease protocol.
 */

import { describe, it, expect, assert } from "vitest";
import {
  completeTaskBody,
  enqueueTaskBody,
  failTaskBody,
  heartbeatTaskBody,
  parseTaskListResponse,
  parseWorkerTask,
  pollTasksBody,
} from "./queues.js";
import { ActeonClient } from "./client.js";
import { HttpError } from "./errors.js";

// ---------------------------------------------------------------------
// Request body builders
// ---------------------------------------------------------------------

describe("queue request body builders", () => {
  it("poll body minimal", () => {
    assert.deepEqual(pollTasksBody("ns", "t1"), { namespace: "ns", tenant: "t1" });
  });

  it("poll body snake-cases optional fields", () => {
    const body = pollTasksBody("ns", "t1", {
      maxTasks: 5,
      leaseSeconds: 120,
      workerId: "worker-7",
    });
    assert.equal(body.max_tasks, 5);
    assert.equal(body.lease_seconds, 120);
    assert.equal(body.worker_id, "worker-7");
  });

  it("enqueue body — payload required, max_attempts optional", () => {
    const minimal = enqueueTaskBody("ns", "t1", "send_email", { to: "x" });
    assert.deepEqual(minimal, {
      namespace: "ns",
      tenant: "t1",
      action_type: "send_email",
      payload: { to: "x" },
    });

    const full = enqueueTaskBody("ns", "t1", "send_email", {}, { maxAttempts: 3 });
    assert.equal(full.max_attempts, 3);
  });

  it("heartbeat body — extend_seconds optional", () => {
    const minimal = heartbeatTaskBody("ns", "t1", "lease-1");
    assert.deepEqual(minimal, {
      namespace: "ns",
      tenant: "t1",
      lease_token: "lease-1",
    });

    const full = heartbeatTaskBody("ns", "t1", "lease-1", 90);
    assert.equal(full.extend_seconds, 90);
  });

  it("complete body carries result verbatim", () => {
    const body = completeTaskBody("ns", "t1", "lease-1", { ok: true });
    assert.deepEqual(body, {
      namespace: "ns",
      tenant: "t1",
      lease_token: "lease-1",
      result: { ok: true },
    });
  });

  it("fail body carries error + retryable", () => {
    const body = failTaskBody("ns", "t1", "lease-1", "boom", false);
    assert.deepEqual(body, {
      namespace: "ns",
      tenant: "t1",
      lease_token: "lease-1",
      error: "boom",
      retryable: false,
    });
  });
});

// ---------------------------------------------------------------------
// Response parsers
// ---------------------------------------------------------------------

function wireTask(
  overrides: Record<string, unknown> = {},
): Record<string, unknown> {
  return {
    task_id: "task-1",
    queue: "billing",
    action_type: "charge",
    payload: { amount: 42 },
    status: "leased",
    attempt: 1,
    max_attempts: 5,
    lease_token: "lease-1",
    lease_expires_at: "2026-06-10T00:01:00Z",
    chain_id: "chain-1",
    workflow_execution_id: "exec-1",
    created_at: "2026-06-10T00:00:00Z",
    updated_at: "2026-06-10T00:00:30Z",
    ...overrides,
  };
}

describe("parseWorkerTask", () => {
  it("parses all fields", () => {
    const task = parseWorkerTask(wireTask({ result: { ok: true }, error: "e" }));
    assert.equal(task.taskId, "task-1");
    assert.equal(task.queue, "billing");
    assert.equal(task.actionType, "charge");
    assert.deepEqual(task.payload, { amount: 42 });
    assert.equal(task.status, "leased");
    assert.equal(task.attempt, 1);
    assert.equal(task.maxAttempts, 5);
    assert.equal(task.leaseToken, "lease-1");
    assert.equal(task.leaseExpiresAt, "2026-06-10T00:01:00Z");
    assert.deepEqual(task.result, { ok: true });
    assert.equal(task.error, "e");
    assert.equal(task.chainId, "chain-1");
    assert.equal(task.workflowExecutionId, "exec-1");
    assert.equal(task.createdAt, "2026-06-10T00:00:00Z");
    assert.equal(task.updatedAt, "2026-06-10T00:00:30Z");
  });

  it("optional fields drop cleanly", () => {
    const task = parseWorkerTask({
      task_id: "task-2",
      queue: "q",
      action_type: "a",
      payload: null,
      status: "pending",
      attempt: 0,
      max_attempts: 1,
      created_at: "2026-06-10T00:00:00Z",
      updated_at: "2026-06-10T00:00:00Z",
    });
    assert.equal(task.leaseToken, undefined);
    assert.equal(task.leaseExpiresAt, undefined);
    assert.equal(task.result, undefined);
    assert.equal(task.error, undefined);
    assert.equal(task.chainId, undefined);
    assert.equal(task.workflowExecutionId, undefined);
  });

  it("parseTaskListResponse unwraps the tasks envelope", () => {
    const tasks = parseTaskListResponse({ tasks: [wireTask(), wireTask({ task_id: "task-2" })] });
    assert.equal(tasks.length, 2);
    assert.equal(tasks[1].taskId, "task-2");

    assert.deepEqual(parseTaskListResponse({}), []);
  });
});

// ---------------------------------------------------------------------
// Client method URL + body behaviour via a captured `fetch`
// ---------------------------------------------------------------------

interface CapturedFetch {
  url: string;
  init: RequestInit | undefined;
}

/** Replace `globalThis.fetch` with a stub that records the call and
 *  returns the supplied response. Returns the captured-calls array
 *  plus a restore function. */
function captureFetch(
  status: number = 200,
  body: unknown = {},
): { calls: CapturedFetch[]; restore: () => void } {
  const calls: CapturedFetch[] = [];
  const original = globalThis.fetch;
  globalThis.fetch = (async (
    input: string | URL | Request,
    init?: RequestInit,
  ): Promise<Response> => {
    calls.push({ url: String(input), init });
    return new Response(JSON.stringify(body), {
      status,
      headers: { "Content-Type": "application/json" },
    });
  }) as typeof globalThis.fetch;
  return {
    calls,
    restore: () => {
      globalThis.fetch = original;
    },
  };
}

function sentBody(call: CapturedFetch): Record<string, unknown> {
  return JSON.parse(call.init?.body as string) as Record<string, unknown>;
}

describe("queue client URLs and bodies", () => {
  it("pollTasks posts the poll body and unwraps tasks", async () => {
    const { calls, restore } = captureFetch(200, { tasks: [wireTask()] });
    try {
      const c = new ActeonClient("http://x");
      const tasks = await c.pollTasks("billing", "ns", "t1", {
        maxTasks: 2,
        leaseSeconds: 30,
        workerId: "w-1",
      });
      expect(calls[0].url).toEqual("http://x/v1/queues/billing/poll");
      expect(calls[0].init?.method).toEqual("POST");
      expect(sentBody(calls[0])).toEqual({
        namespace: "ns",
        tenant: "t1",
        max_tasks: 2,
        lease_seconds: 30,
        worker_id: "w-1",
      });
      expect(tasks.length).toEqual(1);
      expect(tasks[0].taskId).toEqual("task-1");
    } finally {
      restore();
    }
  });

  it("enqueueTask hits the queue tasks URL and parses a 201", async () => {
    const { calls, restore } = captureFetch(201, wireTask({ status: "pending" }));
    try {
      const c = new ActeonClient("http://x");
      const task = await c.enqueueTask("billing", "ns", "t1", "charge", { amount: 42 }, { maxAttempts: 3 });
      expect(calls[0].url).toEqual("http://x/v1/queues/billing/tasks");
      expect(sentBody(calls[0]).max_attempts).toEqual(3);
      expect(task.status).toEqual("pending");
    } finally {
      restore();
    }
  });

  it("heartbeatTask / completeTask / failTask hit the task verb URLs", async () => {
    const { calls, restore } = captureFetch(200, wireTask());
    try {
      const c = new ActeonClient("http://x");
      await c.heartbeatTask("task-1", "ns", "t1", "lease-1", 45);
      await c.completeTask("task-1", "ns", "t1", "lease-1", { ok: true });
      await c.failTask("task-1", "ns", "t1", "lease-1", "boom");
      expect(calls[0].url).toEqual("http://x/v1/queues/tasks/task-1/heartbeat");
      expect(sentBody(calls[0]).extend_seconds).toEqual(45);
      expect(calls[1].url).toEqual("http://x/v1/queues/tasks/task-1/complete");
      expect(sentBody(calls[1]).result).toEqual({ ok: true });
      expect(calls[2].url).toEqual("http://x/v1/queues/tasks/task-1/fail");
      // retryable defaults to true.
      expect(sentBody(calls[2]).retryable).toEqual(true);
    } finally {
      restore();
    }
  });

  it("getTask returns null on 404", async () => {
    const { calls, restore } = captureFetch(404, { error: "task not found" });
    try {
      const c = new ActeonClient("http://x");
      const task = await c.getTask("task-9", "ns", "t1");
      expect(calls[0].url).toEqual(
        "http://x/v1/queues/tasks/task-9?namespace=ns&tenant=t1",
      );
      expect(task).toBeNull();
    } finally {
      restore();
    }
  });

  it("listTasks carries namespace, tenant, and status filters", async () => {
    const { calls, restore } = captureFetch(200, { tasks: [] });
    try {
      const c = new ActeonClient("http://x");
      const tasks = await c.listTasks("billing", "ns", "t1", "failed");
      expect(calls[0].url).toEqual(
        "http://x/v1/queues/billing/tasks?namespace=ns&tenant=t1&status=failed",
      );
      expect(tasks).toEqual([]);
    } finally {
      restore();
    }
  });

  it("surfaces the server error envelope as an HttpError", async () => {
    const { restore } = captureFetch(409, { error: "lease token mismatch" });
    try {
      const c = new ActeonClient("http://x");
      await expect(
        c.completeTask("task-1", "ns", "t1", "stale", null),
      ).rejects.toThrowError(HttpError);
      await expect(
        c.completeTask("task-1", "ns", "t1", "stale", null),
      ).rejects.toThrow(/lease token mismatch/);
    } finally {
      restore();
    }
  });
});
